src/net/ossl.cc

```diff
@@ -1657,9 +1581,181 @@ class session : public enable_shared_from_this<session>, public session_impl {
     session_type _type;
     bool _eof = false;
     bool _shutdown = false;
+
+    friend int bio_write_ex(BIO* b, const char * data, size_t dlen, size_t * written);
```

Comment:
q: Why are these methods friends?

Reply:
Because they need to access private members of the session class